core/src/main/java/org/apache/accumulo/core/fate/zookeeper/ZooCache.java

```diff
@@ -294,41 +296,46 @@ public List<String> getChildren(final String zPath) {
       @Override
       public List<String> run() throws KeeperException, InterruptedException {
 
-        // only read volatile once for consistency
-        ImmutableCacheCopies lic = immutableCache;
-        var cachedChildren = lic.childrenCache.get(zPath);
-        // null is cached as a value, that is the reason for the secondary containsKey check
-        if (cachedChildren != null || lic.childrenCache.containsKey(zPath)) {
-          return cachedChildren;
+        var zcNode = nodeCache.get(zPath);
+        if (zcNode != null && zcNode.childrenSet) {
+          return zcNode.children;
         }
 
-        cacheWriteLock.lock();
         try {
-          List<String> children = childrenCache.get(zPath);
-          // null is cached as a value, that is the reason for the secondary containsKey check
-          if (children != null || lic.childrenCache.containsKey(zPath)) {
-            return children;
-          }
-
-          final ZooKeeper zooKeeper = getZooKeeper();
+          zcNode = nodeCache.compute(zPath, (zp, zcn) -> {
+            // recheck the children now that lock is held on key
+            if (zcn != null && zcn.childrenSet) {
+              return zcn;
+            }
 
-          children = zooKeeper.getChildren(zPath, watcher);
-          if (children != null) {
-            children = List.copyOf(children);
-          }
-          childrenCache.put(zPath, children);
-          immutableCache = new ImmutableCacheCopies(++updateCount, immutableCache, childrenCache);
-          return children;
-        } catch (KeeperException ke) {
-          if (ke.code() != Code.NONODE) {
-            throw ke;
+            try {
+              final ZooKeeper zooKeeper = getZooKeeper();
+              List<String> children = null;
+              // TODO zookeeper may never return null
```

Comment:
What does this TODO mean? Are you suggesting that `ZooKeeper.getChildren` may only return an empty list and never return null?

Reply:
These changes preserve the existing possibly incorrect behavior as I was not sure if it was incorrect or not.

Reply:
Opened #5047

Reply:
removed the TODOs in 2f2a424